SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc

```diff
@@ -2374,3 +2492,130 @@ void SiPixelDigitizerAlgorithm::module_killing_DB(uint32_t detID) {
+        }
+      } else {
+        if (pixdet->subDetector() == GeomDetEnumerators::SubDetector::PixelBarrel ||
+            pixdet->subDetector() == GeomDetEnumerators::SubDetector::P1PXB) {
+          if (lay == 1) {
+            thePixelThresholdInE = theThresholdInE_BPix_L1;
+          } else if (lay == 2) {
+            thePixelThresholdInE = theThresholdInE_BPix_L2;
+          } else {
+            thePixelThresholdInE = theThresholdInE_BPix;  // no smearing
+          }
+        }
+      }
+
+    } else if (pixdet->type().isTrackerPixel()) {  // Forward disks modules
+
+      if (addThresholdSmearing) {
+        thePixelThresholdInE =
+            CLHEP::RandGaussQ::shoot(engine, theThresholdInE_FPix, theThresholdSmearing_FPix);  // gaussian smearing
+      } else {
+        thePixelThresholdInE = theThresholdInE_FPix;  // no smearing
+      }
+
+    } else {
+      throw cms::Exception("NotAPixelGeomDetUnit") << "Not a pixel geomdet unit" << detID;
+    }
+  }
+
+  // loop on the SimHit extra info class
+  // apply the reweighting for that SimHit on a cluster way
+  bool reweighted = false;
+  std::vector<PixelSimHitExtraInfo>::iterator loopTempSH;
+  for (loopTempSH = newClass_Sim_extra.begin(); loopTempSH != newClass_Sim_extra.end(); ++loopTempSH) {
+    signal_map_type theDigiSignal;
+    PixelSimHitExtraInfo TheNewInfo = *loopTempSH;
+    reweighted = TheNewSiPixelChargeReweightingAlgorithmClass->lateSignalReweight(
+        pixdet, digis, TheNewInfo, theDigiSignal, tTopo, engine);
+    if (!reweighted) {
+      // loop on the non-reweighthed digis associated to the considered SimHit
+      std::vector<PixelDigi>::const_iterator loopDigi;
+      for (loopDigi = digis.begin(); loopDigi != digis.end(); ++loopDigi) {
+        unsigned int chan = loopDigi->channel();
+        // check if that digi is related to the SimHit
+        if (loopTempSH->isInTheList(chan)) {
+          float corresponding_charge = loopDigi->adc();
+          theDigiSignal[chan] += Amplitude(corresponding_charge, corresponding_charge);
+        }
+      }
+    }
+
+    // transform theDigiSignal into digis
+    int Thresh_inADC = int(thePixelThresholdInE / theElectronPerADC);
+    for (signal_map_const_iterator i = theDigiSignal.begin(); i != theDigiSignal.end(); ++i) {
+      float signalInADC = (*i).second;  // signal in ADC
+      if (signalInADC > 0.) {
+        if (signalInADC >= Thresh_inADC) {
+          int chan = (*i).first;  // channel number
+          std::pair<int, int> ip = PixelDigi::channelToPixel(chan);
+          int adc = int(signalInADC);
+          // add MissCalibration
+          if (doMissCalInLateCR) {
+            int row = ip.first;
+            int col = ip.second;
+            adc =
+                int(missCalibrate(detID, tTopo, pixdet, col, row, signalInADC * theElectronPerADC));  //full misscalib.
+          }
+
+          if (signalInADC > theAdcFullScLateCR)
+            adc = theAdcFullScLateCR;  // Check maximum value
```

Comment:
```suggestion
          if (signalInADC > theAdcFullScLateCR)
            adc = theAdcFullScLateCR;  // Check maximum value
          else {
            int chan = (*i).first;  // channel number
            std::pair<int, int> ip = PixelDigi::channelToPixel(chan);
            int adc = int(signalInADC);
            // add MissCalibration
            if (doMissCalInLateCR) {
              int row = ip.first;
              int col = ip.second;
              adc = int(missCalibrate(detID, tTopo, pixdet, col, row, signalInADC * theElectronPerADC));
            }
          }
```

Reply:
@perrotta I don't understand your last suggestion. 

At the moment, if I am correct, this missCalibrate function will reduce the large values of adc signals. So in principle we are safe with checking the max value first. I suppose no one will change this function… Except if at some point this code is also used for upgrade and people have something else in mind…

But I don’t understand what we gain with this kind of change… 

The line  
`    std::pair<int, int> ip = PixelDigi::channelToPixel(chan); `
is any way needed when we fill the new Digi later
`    New_digis.emplace_back(ip.first, ip.second, adc);  `
so it can not be within the if sequence you propose...

Reply:
Yes, the exact code suggestion of mine was indeed incorrect: I updated it to define int adc outside the if block. As such it looks correct to me, but please check.

The idea is: if `signalInADC > theAdcFullScLateCR` then it is a waste of cpu resources to calculate `adc` from line 2592 to line 2602, as it would be overwritten anyhow. And `ip` is only needed for that calculation, also being computed uselessly if `signalInADC > theAdcFullScLateCR`

Reply:
Even more compact:
```C++
          int adc = int(signalInADC);
          if (adc > theAdcFullScLateCR) {       // check maximum value
             adc = theAdcFullScLateCR;
          }
          else  if (doMissCalInLateCR) {          // add MissCalibration
            int chan = (*i).first;  // channel number
            std::pair<int, int> ip = PixelDigi::channelToPixel(chan);
            int row = ip.first;
            int col = ip.second;
            adc = int(missCalibrate(detID, tTopo, pixdet, col, row, signalInADC * theElectronPerADC));
          }
```

Reply:
As said,`ip` is also used  [just after](https://github.com/cms-sw/cmssw/blob/e6f4f056eb8b02798cc5f3a3c6c5b224bd1b6d57/SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc#L2613).

In fact, as the `missCalibrate` function will reduce the signalInADC value, we should first computes the effect of the mis-calibration and then check if it is > the max. Because it could happen that first we are at 257, then we miscalibrate and it ends up at 252 which is below the max : 255. If you inverse the sequence, you will check if it is below the max value and decide it should be 255...  So, I think, it is better not to follow your suggestion.

Reply:
If I did not mistakes, the code above reproduces exactly what you have in your PR, with a few less consumed cycles
Still I have two remaining doubts:
- Shouldn't `if (adc > theAdcFullScLateCR)` be `if (adc >= theAdcFullScLateCR)` instead?
- Is it correct to check that upper limit before `missCalibrate`? Shouldn't it be checked after the recalculated `adc` value afterwards?

Reply:
> As said,`ip` is also used [just after](https://github.com/cms-sw/cmssw/blob/e6f4f056eb8b02798cc5f3a3c6c5b224bd1b6d57/SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc#L2613).

Ah, correct!
Yes, I missed it: then the lines that computes it should be moved outside the if block.

> In fact, as the `missCalibrate` function will reduce the signalInADC value, we should first computes the effect of the mis-calibration and then check if it is > the max. Because it could happen that first we are at 257, then we miscalibrate and it ends up at 252 which is below the max : 255. If you inverse the sequence, you will check if it is below the max value and decide it should be 255... So, I think, it is better not to follow your suggestion.

The `missCalibrate` function does not reduce the `signalInADC` value, which remains untouched, but it is the `adc` value which goes below `int(signalInADC)`. That why I posted mu doubt above: "Shouldn't it be checked after the recalculated adc value afterwards?". Please notice that in your original code (as well as in my "translation") you checked the unmodified `signalInADC`, not the rescaled `adc`?

Reply:
>     * Is it correct to check that upper limit before `missCalibrate`? Shouldn't it be checked after the recalculated `adc` value afterwards?

I  think that this check has to be done after `missCalibrate`, as it is originally in the PR. (see the previous example)
 
>     * Shouldn't `if (adc > theAdcFullScLateCR)` be `if (adc >= theAdcFullScLateCR)` instead?

then if we keep the original way of the PR, first `missCalibrate` then `if (adc > theAdcFullScLateCR)`, then we don't care if it is `>= ` or only `>`.

So I am in favor of my original proposal.

Reply:
@perrotta : I write my answer here (at the bottom of the thread), because it will become very difficult to read it later...
 
> The `missCalibrate` function does not reduce the `signalInADC` value, which remains untouched, but it is the `adc` value which goes below `int(signalInADC)`. That why I posted mu doubt above: "Shouldn't it be checked after the recalculated adc value afterwards?". Please notice that in your original code (as well as in my "translation") you checked the unmodified `signalInADC`, not the rescaled `adc`.

You are right. Thanks! I miss that. I will change the code. 

But for the rest of the organisation of that part. I really think it is better to only check the `adc` value after `missCalibrate`, as `missCalibrate` will change this value. Do you agree ?

Reply:
> But for the rest of the organisation of that part. I really think it is better to only check the `adc` value after `missCalibrate`, as `missCalibrate` will change this value. Do you agree ?

Yes, I agree. My comment was only intended to rewrite in a more efficient way what you coded. Of course, if the original logic was wrong then the suggestion to rewrite it differently does not apply any more.

Reply:
`      if (signalInADC > 0.) {
        if (signalInADC >= Thresh_inADC) {
          int chan = (*i).first;  // channel number
          std::pair<int, int> ip = PixelDigi::channelToPixel(chan);
          int adc = int(signalInADC);
          // add MissCalibration
          if (doMissCalInLateCR) {
            int row = ip.first;
            int col = ip.second;
            adc =
                int(missCalibrate(detID, tTopo, pixdet, col, row, signalInADC * theElectronPerADC));  //full misscalib.
          }

          if (adc > theAdcFullScLateCR)
            adc = theAdcFullScLateCR;  // Check maximum value

#ifdef TP_DEBUG
          LogDebug("Pixel Digitizer") << (*i).first << " " << (*i).second << " " << signalInADC << " " << adc
                                      << ip.first << " " << ip.second;
#endif

          // Load digis
          New_digis.emplace_back(ip.first, ip.second, adc);
        }
      }
`

ok, maybe here I could remove the first if...